crates/wit-parser/src/sizealign.rs

```diff
@@ -116,64 +292,249 @@ impl SizeAlign {
+        ElementInfo::new(align_to_arch(size_sum, align), align)
     }
 }
 
-fn int_size_align(i: Int) -> (usize, usize) {
+fn int_size_align(i: Int) -> ElementInfo {
     match i {
-        Int::U8 => (1, 1),
-        Int::U16 => (2, 2),
-        Int::U32 => (4, 4),
-        Int::U64 => (8, 8),
+        Int::U8 => Alignment::Bytes(NonZeroUsize::new(1).unwrap()),
+        Int::U16 => Alignment::Bytes(NonZeroUsize::new(2).unwrap()),
+        Int::U32 => Alignment::Bytes(NonZeroUsize::new(4).unwrap()),
+        Int::U64 => Alignment::Bytes(NonZeroUsize::new(8).unwrap()),
     }
+    .into()
 }
 
+/// Increase `val` to a multiple of `align`;
+/// `align` must be a power of two
 pub(crate) fn align_to(val: usize, align: usize) -> usize {
     (val + align - 1) & !(align - 1)
 }
+
+/// Increase `val` to a multiple of `align`, with special handling for pointers;
+/// `align` must be a power of two or `Alignment::Pointer`
+pub fn align_to_arch(val: ArchitectureSize, align: Alignment) -> ArchitectureSize {
+    match align {
+        Alignment::Pointer => {
+            let new32 = align_to(val.bytes, 4);
+            let new64 = align_to(val.bytes + val.add_for_64bit, 8);
+            ArchitectureSize::new(new32, new64 - new32)
+        }
+        Alignment::Bytes(align_bytes) => {
+            let new32 = align_to(val.bytes, align_bytes.get());
+            let new64 = align_to(val.bytes + val.add_for_64bit, align_bytes.get());
+            ArchitectureSize::new(new32, new64 - new32)
+        }
+    }
+}
+
+/// Compatibility with older versions:
+/// Collect size and alignment for sub-elements of a structure, simpler interface only supporting wasm32
+#[derive(Default)]
+pub struct SizeAlign32(SizeAlign64);
+
+impl SizeAlign32 {
+    pub fn new() -> Self {
+        Default::default()
+    }
+    pub fn fill(&mut self, resolve: &Resolve) {
+        self.0.fill(resolve);
+    }
+    pub fn size(&self, ty: &Type) -> usize {
+        self.0.size(ty).size_wasm32()
+    }
+    pub fn align(&self, ty: &Type) -> usize {
+        self.0.align(ty).align_wasm32()
+    }
+    pub fn field_offsets<'a>(
+        &self,
+        types: impl IntoIterator<Item = &'a Type>,
+    ) -> Vec<(usize, &'a Type)> {
+        self.0
+            .field_offsets(types)
+            .drain(..)
+            .map(|(offs, ty)| (offs.size_wasm32(), ty))
+            .collect()
+    }
+    pub fn payload_offset<'a>(
+        &self,
+        tag: Int,
+        cases: impl IntoIterator<Item = Option<&'a Type>>,
+    ) -> usize {
+        self.0.payload_offset(tag, cases).size_wasm32()
+    }
+    pub fn record<'a>(&self, types: impl Iterator<Item = &'a Type>) -> (usize, usize) {
+        let info = self.0.record(types);
+        (info.size.size_wasm32(), info.align.align_wasm32())
+    }
+    pub fn params<'a>(&self, types: impl IntoIterator<Item = &'a Type>) -> (usize, usize) {
+        self.record(types.into_iter())
+    }
+}
+
+/// compatibility alias
+pub type SizeAlign = SizeAlign32;
```

Comment:
Would it be possible to go ahead and remove these typedefs/aliases? I think it'd be best to only have the new `SizeAlign64` (probably renamed to `SizeAlign` if it's the only one) which would force users to update one way or another. There's enough conveniences on the above types that not using 64-bit support I think would be reasonable.

Reply:
I implemented this fully compatible drop-in replacement type to enable migrating wit-bindgen (and perhaps other users of `SizeAlign`) more gradually to the new API - to prevent locking a new version of each crate to each other. Once this MR merges wit-bindgen would require many code changes directly when updating its wasm-tools dependency.

But as you seem willing to take the risk (and are the maintainer of both repositories) … requiring migration clearly is the more clean solution in the long run.

Reply:
Personally I think it's ok to go ahead and replace the old `SizeAlign`. If necessary `wit-bindgen-core` can contain the `SizeAlign32` helper you have written here and then the code generators can be temporarily updated to using that.